ppb/features/collision.py

```diff
@@ -0,0 +1,108 @@
+"""
+from itertools import product
+
+from ppb.sprites import BaseSprite
+from ppb.systemslib import System
+
+
+class CollidedWithMixin:
+    """
+    Provides a mechanism for CollisionCheckerSystem to inform sprites when
+    they collide with something.
+
+    This is a mixin added to CanCollideCircleMixin and
+    CanCollideSquareMixin. You do not need to to add it to your sprites if
+    you are using one of them.
+    """
+
+    def collided_with(self, other):
+        """
+        Another sprite has collided with this sprite.
+
+        This is meant to be overridden in your sprites. If a sprite has a
+        collider mixin, but doesn't respond to those collisions, you can
+        ignore this.
+        """
+
+
+class CanCollideCircleMixin(BaseSprite, CollidedWithMixin):
+    """
+    Defines a circular region collider and provides a method to check for
+    collision.
+    """
+    radius: float = 1
+
+    def collides_with(self, other: 'CanCollideCircleMixin'):
+        """
+        Check if this sprite collides with another sprite.
+
+        Both sprites must be CanCollideCircleMixin sprites.
+        """
+        distance = (self.position - other.position).length
+        collide_distance = self.radius + other.radius
+        return distance <= collide_distance
+
+
+class CanCollideSquareMixin(BaseSprite, CollidedWithMixin):
+    """
+    Uses the defined square on a sprite to provid a method to check for
+    collision.
+    """
+
+    def collides_with(self, other: 'CanCollideSquareMixin'):
+        """
+        Check if this sprite collides with another sprite.
+
+        Both sprites must be CanCollideSquareMixin sprites.
+        """
+        furthest_left = min(self.left, other.left)
+        furthest_right = max(self.right, other.right)
+        furthest_up = max(self.top, other.top)
+        furthest_down = min(self.bottom, other.bottom)
+        collide_distance = self.size + other.size
+        return all([furthest_right - furthest_left <= collide_distance,
+                    furthest_up - furthest_down <= collide_distance])
+
+
+class CollisionCheckerSystem(System):
+    """
+    A subsystem to check for collisions after the update step of the system.
+
+    To use, add CollisionCheckerSystem to the list of systems in ppb.run
+    or the constructor of ppb.GameEngine.
+
+    You should only use one type of collider with this system. It is not
+    designed to check fallback collision types. It is allowed, however, to
+    have some sprites with colliders and some without.
+    """
+    primed = False
+
+    def on_update(self, event, signal):
+        self.primed = True
+
+    def on_idle(self, event, signal):
+        if not self.primed:
+            return
+        for s1, s2 in product(event.scene, event.scene):
+            if s1 is s2:
+                continue
+            try:
+                if s1.collides_with(s2):
+                    s1.collided_with(s2)
+                    s2.collided_with(s1)
```

Comment:
This line will mean that `.collided_with()` will be called twice.

Reply:
. . . yes. From two different objects, who are likely to have two different collided_with methods.

Reply:
No, I mean it'll detect twice.

First as s1 as Spam and s2 as Foo, and then again as s1 as Foo and s2 as Spam.

Reply:
No.

`s1.collides_with(s2)` == check for collision.

`s1.collided_with(s2)` == s1 responds to a collision detected with s2.
`s2.collided_with(s1)` == s2 responds to a collision detected with s1.

Reply:
Just making sure, you mean the product versus combination problem (from the top this block?)

Reply:
Yup. Product is probably what you want for asymmetric collisions, though.